### ISSUE TITLE
Should the Radio Group be usable without a Label?

### ISSUE THREAD

**Issue by amcgee** [dhis2/ui-core#457]
Right now, when using `RadioGroup` without a `label` prop, the component creates an empty `<span>` which messes up the alignment of the radio buttons.  Below is a screenshot for context.  I was able to work around this by setting `display: none` on the span in CSS, but maybe the `label` prop shouldn't be required and the `<span>` should only render if it's specified?  Or maybe I'm breaking some design rules I shouldn't be here @cooper-joe?

I think in this case if we want to require a label, we should be able to render the label to the left of the radio group, rather than above it.

<img width="1232" alt="Screenshot 2019-09-27 10 56 40" src="https://user-images.githubusercontent.com/947888/65757408-56e73080-e117-11e9-9400-30f8bbb83529.png">

And this is what it looks like without the CSS hack (with no label specified):

<img width="565" alt="Screenshot 2019-09-27 11 11 37" src="https://user-images.githubusercontent.com/947888/65757527-9e6dbc80-e117-11e9-835f-2bfd0010479a.png">

**Comment by cooper-joe** [comment]
Interesting use case, thanks for raising this @amcgee.

I can see here how a label-less radioGroup is useful. Normally my advice would be: "include a label, otherwise how will the user know what the radio buttons apply to?". However, in this use case it is somewhat self-evident that the radio buttons apply to the type of execution. For beginner users it may still be useful to have a label of 'Type' or something similar. 

I do see how `labelPosition` with options `top` and `left` could be useful, especially in these compact situations. I would imagine this would be relatively painless to implement, so I think it may be worth including to boost the flexibility of these field components. What do you think @dhis2/team-apps?

In conclusion:
- I am in favour of adding a labelPosition property
- I do agree that if a label is not included the `span` should be omitted, although labels are highly recommended.

**Comment by Mohammer5** [comment]
We're in the process of changing `ui-core` right now (see the `next` branch).
We're extracting input components from field components (field = [label +] input [+ help] [+error]),
so the component you need could be composed easily with the atoms `ui-core` will provide.

It shouldn't be too much of a problem then to create a `RadioGroupField` that can handle multiple label positions, which will require custom css for the molecule (and the current rule is: no css for anything but atoms).

@varl here I am again, proposing to get rid of this rule because molecules will need "glue" styles until we have figured out a proper way (maybe layout components, maybe not).

If we stick to that rule, it will still be fairly easy to create this component in apps/libs because all relevant atoms will be exposed.

**Comment by varl** [comment]
> @varl here I am again, proposing to get rid of this rule because molecules will need "glue" styles until we have figured out a proper way (maybe layout components, maybe not).

Too make it easier for `ui-forms` we must stick to that rule for the `*-Field` components. `ui-forms` cannot provide its own glue.

> I do see how `labelPosition` with options `top` and `left` could be useful, especially in these compact situations. I would imagine this would be relatively painless to implement, so I think it may be worth including to boost the flexibility of these field components

Are we talking about `Label`s in general? As in for `InputField`, `SelectField`, etc.? Then I agree that it is something we could look into.

**Comment by cooper-joe** [comment]
> Are we talking about `Label`s in general? As in for `InputField`, `SelectField`, etc.? Then I agree that it is something we could look into. I am wary of the complexity this could introduce though.

Yes, I think consistency between all `Field` composed elements is a good thing, so I think positioning options would apply to all composed `Fields`. I am also quite wary of starting down this road and ending up at 'options overload'. I do see immediate use cases for left-aligned `labels` though (Capture/Tracker dashboard will benefit immediately), so the trade-off is potentially worth it.

**Comment by varl** [comment]
@cooper-joe Ok, acknowledged.

One question, is this in addition to the "Inline label" that's specced in: https://github.com/dhis2/design-system/blob/master/molecules/select.md#inline-label?

The "inline label" is not something that is going to extend to e.g. the `Input`?

**Comment by cooper-joe** [comment]
> @cooper-joe Ok, acknowledged.
> 
> Is this in addition to the "Inline label" that's specced in: https://github.com/dhis2/design-system/blob/master/molecules/select.md#inline-label?
> 
> And, could the "inline label" extend to the `InputField` in the future?

Good question. We're moving away from the original issue here, so I've opened a new issue to continue the discussion https://github.com/dhis2/ui-core/issues/460

**Comment by varl** [comment]
Answer to the topic question:

Technically you should be able to use the `Radio` component without the label. This will be refactored in `next` (#430) because the entire `RadioGroup` implementation in `master` is wrong and doesn't match the rest of the API in `ui-core`.

**Comment by HendrikThePendric** [comment]
> the entire `RadioGroup` implementation in `master` is wrong and doesn't match the rest of the API in `ui-core`

Yes, the current implementation pre-dates the `Field` approach that emerged recently. I'll start refactoring this once I finish the `TextArea`

**Comment by Mohammer5** [comment]
> Too make it easier for ui-forms we must stick to that rule for the *-Field components. ui-forms cannot provide its own glue so it would never be able to compose variants of the *-Field components to match any unsupported needs it has that we do not want in ui-core.

@varl I don't think that the RadioGroup belongs into `ui-forms`, it should be in `ui-core`, so I'm referring back to what I said (please consider it applied to `ui-core`, not `ui-forms` which should have no styles whatsoever):

> @varl here I am again, proposing to get rid of this rule because molecules will need "glue" styles until we have figured out a proper way (maybe layout components, maybe not).

If a label can be either on the top or on the left, we either have two molecules, layout components or glue in the `ui-core` molecules. If we really want to avoid any glue, we need to have two molecules (`RadioGroup` with label on the top & `RadioGroup` with lavel on the left), right?

**Comment by amcgee** [comment]
For me the simplest option here is just to add `{label && <span>{label}</span>}` (simplified) so that the span doesn't render if no label is passed - that way the consumer is free to put the label somewhere else or omit it.  I understand if we want to lock down the relative position of the label to "left" or "top", but perhaps that would be a simple interim solution?

**Comment by Mohammer5** [comment]
Once we have a RadioGroup in `ui-core`, this will def. be the case, we're doing this with all other fields as well

**Comment by amcgee** [comment]
... but isn’t it currently in ui-core?  That’s where I used it and opened this issue...

**Comment by varl** [comment]
@amcgee There is a `RadioGroup` in `master`, but not in `next`. The `RadioGroup` in `master` has the wrong abstraction level and doesn't match the other form controls.

The discussion here deals with multiple issues, so i will push a fix (#475) to close this issue and avoid confusion about `next` and `master`.
